### PR TITLE
8305563: [vectorapi]: Initial aarch64 backend implementation for FP16 operations

### PR DESCRIPTION
**Pull request by Bhavana-Kilambi**
This patch adds aarch64 (Neon and SVE) backend implementation for the following FP16 operations for which C2 support was added in the commit https://github.com/openjdk/panama-vector/commit/7460d9327aac7d1d2ba6aed4e7137a417dbf6a97 -

Unary operations : AbsVHF, NegVHF
Binary operations : AddVHF, SubVHF, MulVHF, DivVHF
Ternary operations : AddReductionVHF, FmaVHF
Conversion operations : VectorCastHF2D, VectorCastD2HF

Some of these operations on Neon machines, require features "fphp" and "asimdhp" to be available. Feature detection for these features have also been added.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8305563](https://bugs.openjdk.org/browse/JDK-8305563): [vectorapi]: Initial aarch64 backend implementation for FP16 operations (**Enhancement** - P4)

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - no project role)
 * [Xiaohong Gong](https://openjdk.org/census#xgong) (@XiaohongGong - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-vector.git pull/217/head:pull/217` \
`$ git checkout pull/217`

Update a local copy of the PR: \
`$ git checkout pull/217` \
`$ git pull https://git.openjdk.org/panama-vector.git pull/217/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 217`

View PR using the GUI difftool: \
`$ git pr show -t 217`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-vector/pull/217.diff">https://git.openjdk.org/panama-vector/pull/217.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/panama-vector/pull/217#issuecomment-1607388861)

### CODE REVIEW

File: src/hotspot/cpu/aarch64/aarch64_vector.ad [hunk 1]
```diff
@@ -226,6 +226,20 @@ source %{
...
+      case Op_AddReductionVHF:
+        // FEAT_FP16 is enabled if both "fphp" and "asimdhp" features are supported.
+        // Only the Neon instructions need this check. SVE supports 16-bit floats by default.
+        if (UseSVE == 0 && (!VM_Version::supports_fphp() || !VM_Version::supports_asimdhp())) {
+          return false;
+        }
+        break;
       default:
         break;
     }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_vector.ad:245]:
```suggestion
        if (UseSVE > 0 || (VM_Version::supports_fphp() && VM_Version::supports_asimdhp())) {
         break;
        }
        return false;
      default:
        break;
    }
```

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 2]
```diff
@@ -2707,25 +2718,29 @@ template<typename R, typename... Rx>
...
+#define INSN(NAME, op1, op2, op3)                                                           \
+  void NAME(FloatRegister Vd, SIMD_Arrangement T, FloatRegister Vn, FloatRegister Vm) {     \
+    starti;                                                                                 \
+    assert(T== T4H || T == T8H || T == T2S || T == T4S || T == T2D, "invalid arrangement"); \
+    int op22 = (T == T2S || T == T4S) ? 0b0 : 0b1;                                          \
+    int op21 = (T == T4H || T == T8H) ? 0b0 : 0b1;                                          \
+    int op15 = (T == T4H || T == T8H) ? 0b00 : 0b11;                                        \
+    f(0, 31), f((int)T & 1, 30), f(op1, 29), f(0b01110, 28, 24), f(op2, 23);                \
+    f(op22, 22); f(op21, 21), rf(Vm, 16), f(op15, 15, 14), f(op3, 13, 10), rf(Vn, 5);       \
+    rf(Vd, 0);                                                                              \
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:2730]: This is a bit too big for an inline function. Please consider taking the body of this asm and putting it into assembler_aarch64.cpp. See `Assembler::add_sub_immediate` for an example of how this should be done.

**Review by theRealAph — COMMENTED** [review]
This all looks reasonable enough, except for a couple of minor niggles.

**Comment by Bhavana-Kilambi** [comment]
@theRealAph Thank you for your review. Will update a new patchset soon.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 3]
```diff
@@ -2706,26 +2717,28 @@ template<typename R, typename... Rx>
 
 #undef INSN
 
+  void fp_arithmetic(Instruction_aarch64 &current_insn, FloatRegister Vd,
+                     SIMD_Arrangement T, FloatRegister Vn, FloatRegister Vm,
+                     int op1, int op2, int op3);
+
 // Advanced SIMD three same
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:2724]:
```suggestion
    // Advanced SIMD three same

    void adv_simd_three_same(Instruction_aarch64 &current_insn, FloatRegister Vd,
                     SIMD_Arrangement T, FloatRegister Vn, FloatRegister Vm,
                     int op1, int op2, int op3);

```
Always match names with names in the Architecture Reference Manual where we can.

**Review by XiaohongGong — APPROVED** [review]
LGTM!

**Review by theRealAph — APPROVED** [review]
Looks great. Thanks.

**Pull request closed without merging**